specification/dartLangSpec.tex

```diff
@@ -21253,39 +21261,2076 @@ \subsubsection{Informal Subtype Rule Descriptions}
+
+The normalization which is described here will preserve
+distinctions like the former (such as \DYNAMIC{} and \code{Object?}),
+and it will eliminate distinctions like the latter
+(such as \code{Never} and $X$).
+
+In particular, \SubtypeNE{S}{T} and \SubtypeNE{T}{S} holds if and only if
+\NormalizedTypeOf{$T$} has the same canonical syntax as \NormalizedTypeOf{$S$}
+(\ref{standardUpperBoundsAndStandardLowerBounds}),
+modulo replacement of atomic top types
+(e.g., \code{List<C<\DYNAMIC>{}>} and \code{List<myPrefix.C<\VOID>{}>}).%
+}
+
+\LMHash{}%
+The function \Index{\NormalizedTypeOfName} is then defined as follows:
+\BlindDefineSymbol{T_a, T_u, T_r}%
+Let $T_a$ be a type
+(\commentary{where `a' stands for `argument'}).
+Let $T_u$ be the transitive alias expansion of $T_a$
+(\commentary{`u' for `unaliased'}, \ref{typedef}).
+Then \DefEquals{\NormalizedTypeOf{$T_a$}}{T_r}
+(\commentary{`r' for result}),
+where $T_r$ is determined as follows:
+
+\begin{itemize}
+\item
+  If $T_u$ is atomic then $T_r$ is $T_u$.
+\item If $T_u$ is of the form \code{FutureOr<$T$>}
+  then let $S$ be \NormalizedTypeOf{$T$} and then:
+  \begin{itemize}
+  \item If $S$ is a top type or \code{Object} then $T_r$ is $S$.
+  \item If $S$ is \code{Never} then $T_r$ is \code{Future<Never>}.
+  \item If $S$ is \code{Null} then $T_r$ is \code{Future<Null>?}.
+  \item Otherwise, $T_r$ is \code{FutureOr<S>}.
+  \end{itemize}
+\item If $T_u$ is of the form \code{$T$?}\ then
+  let $S$ be \NormalizedTypeOf{$T$} and then:
+  \begin{itemize}
+  \item If $S$ is a top type then $T_r$ is $S$.
+  \item If $S$ is \code{Never} or \code{Null} then
+    $T_r$ is \code{Null}.
+  \item If $S$ is \code{FutureOr<$R$>} where $R$ is nullable then
+    $T_r$ is $S$.
+  \item If $S$ is \code{$R$?}\ for some $R$ then $T_r$ is \code{$R$?}.
+  \item Otherwise, $T_r$ is \code{$S$?}.
+  \end{itemize}
+\item If $T_u$ is a type variable $X$ with bound $B$ then:
+  \begin{itemize}
+  \item If $B$ is \code{Never} then $T_r$ is \code{Never}.
+  \item If $B$ is a type variable $Y$
+    and \NormalizedTypeOf{$Y$} is \code{Never}
+    then $T_r$ is \code{Never}.
+  \item Otherwise, $T_r$ is $X$.
+  \end{itemize}
+\item If $T_u$ is of the form \code{$X$\,\&\,$T$}
+  where $X$ is a type variable with bound $B$
+  then let $S$ be \NormalizedTypeOf{$T$} and then:
+  \begin{itemize}
+  \item If $S$ is \code{Never} then $T_r$ is \code{Never}.
+  \item If $S$ is $X$ or a top type then $T_r$ is $X$.
+  \item If \SubtypeNE{\NormalizedTypeOf{$B$}}{S} then $T_r$ is $X$.
+  \item Otherwise, $T_r$ is \code{$X$\,\&\,$S$}.
+  \end{itemize}
+\item If $T_u$ is of the form \code{$C$<\List{T}{1}{k}>}
+  then $T_r$ is \code{$C$<\List{R}{1}{k}>},
+  where $R_i$ is \NormalizedTypeOf{$T_i$}, for $i \in 1 .. k$.
+\item If $T_u$ is of the form
+  \FunctionTypePositionalStd{T_0}
+
+  \noindent
+  then $T_r$ is
+  \FunctionTypePositional{R_0}{ }{X}{B}{s}{R}{n}{k}
```

Comment:
The feature spec seem to require the normalization of the bounds of type variables too. Is it omitted on purpose?